# Book Recomendataion System

In [82]:
#importing the libraries.
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sn
%matplotlib inline

In [83]:
# import the dataset.
books = pd.read_csv('Books.csv')
ratings = pd.read_csv('Ratings.csv')
users = pd.read_csv('Users.csv')

C:\Users\HP\AppData\Local\Temp\ipykernel_3400\2931296443.py:2: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv('Books.csv')


In [84]:
books.shape,ratings.shape,users.shape

((271360, 8), (1149780, 3), (278858, 3))

In [85]:
# checking the nulll values.
books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [86]:
books.head(2)

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...

In [87]:
ratings.head(2)

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5

In [88]:
users.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

In [89]:
ratings.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

In [90]:
# duplicate values.
books.duplicated().sum()

0

In [91]:
users.duplicated().sum()

0

In [92]:
ratings.duplicated().sum()

0

# EDA

# Popularity based recommender system

In [93]:
# we will consider only those books which got minimum of 250 votes.
# for that we will merge the rating and books dataset on top of ISBN column

In [94]:
# merging ratings and books dataset on top of 'ISBN' columns
ratings_with_book  = ratings.merge(books, on='ISBN')

books: ((271360, 8),ratings: (1149780, 3)
but the shape of merged books and rating is less than 1149780, bcoz some rated books were not in the book columns

In [95]:
ratings_with_book.head(1)

User-ID        ISBN  Book-Rating            Book-Title Book-Author  \
0   276725  034545104X            0  Flesh Tones: A Novel  M. J. Rose   

  Year-Of-Publication         Publisher  \
0                2002  Ballantine Books   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/034545104X.0...

In [96]:
ratings_with_book.shape

(1031136, 10)

In [97]:
# now we will select only those books whose average ratings are >250\
# we will apply groupby operation on book-title.
# we can't apply groupby on ISBN no, since same book(some) has different ISBN number.

In [98]:
# ratings_with_book.groupby('Book-Title').count()['Book-Rating'] is coming to be the series of books
# a series basically contains only one column , so it must be converted into dataframe to get multiple feature columns
# reset_index(), is useful when the index needs to be treated as a column
num_rating_df = ratings_with_book.groupby('Book-Title').count()['Book-Rating'].reset_index()

In [99]:
# renaming the Book-Rating column by 'num_ratings'
num_rating_df.rename(columns = {'Book-Rating':'num_ratings'}, inplace = True)

In [100]:
# num_ratings is the number of ratings/or total votes given to a book
num_rating_df.head(2)

Book-Title  num_ratings
0   A Light in the Storm: The Civil War Diary of ...            4
1                              Always Have Popsicles            1

In [101]:
# averge rating of each book.
avg_rating_df = ratings_with_book.groupby('Book-Title').mean()['Book-Rating'].reset_index()
avg_rating_df.rename(columns = {'Book-Rating':'avg_ratings'}, inplace = True)

In [102]:
avg_rating_df.head(2)

Book-Title  avg_ratings
0   A Light in the Storm: The Civil War Diary of ...         2.25
1                              Always Have Popsicles         0.00

In [103]:
# merging the average and num-rating columns, since we want unique book to have a num-rating and an average-rating.

In [104]:
popular_df = num_rating_df.merge(avg_rating_df, on='Book-Title')
popular_df

Book-Title  num_ratings  \
0        A Light in the Storm: The Civil War Diary of ...            4   
1                                   Always Have Popsicles            1   
2                    Apple Magic (The Collector's series)            1   
3        Ask Lily (Young Women of Faith: Lily Series, ...            1   
4        Beyond IBM: Leadership Marketing and Finance ...            1   
...                                                   ...          ...   
241066                                      Ã?Â?lpiraten.            2   
241067                     Ã?Â?rger mit Produkt X. Roman.            4   
241068                                Ã?Â?sterlich leben.            1   
241069                              Ã?Â?stlich der Berge.            3   
241070                                  Ã?Â?thique en toc            2   

        avg_ratings  
0          2.250000  
1          0.000000  
2          0.000000  
3          8.000000  
4          0.000000  
...             ...  
241066     0.000000  
241067     5.250000  
241068     7.000000  
241069     2.666667  
241070     4.000000  

[241071 rows x 3 columns]

In [105]:
# now we will select only those books which received number of ratings > 250
# took all books whose no of ratings>250, sorted them according to 'avg_rating.'
# finally selecting top 50 avg rated books
# these numbers are selected based on experiments.
popular_df = popular_df[popular_df['num_ratings'] > 250].sort_values('avg_ratings', ascending=False).head(50)

In [106]:
# now we need to add author name, publisher, and the image url to these 50 selected rated books.
# so we need to use merge function, which will act as intersection.

In [107]:
 # popular_df.merge(books, on = 'Book-Author')
###-- this was giving error, since we can only merge on common columns, here Author is not present in the popular_df.

In [108]:
# so, 
popular_df.merge(books, on = 'Book-Title')

Book-Title  num_ratings  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)          428   
1    Harry Potter and the Prisoner of Azkaban (Book 3)          428   
2    Harry Potter and the Prisoner of Azkaban (Book 3)          428   
3         Harry Potter and the Goblet of Fire (Book 4)          387   
4         Harry Potter and the Goblet of Fire (Book 4)          387   
..                                                 ...          ...   
191                                  Lord of the Flies          259   
192                            A Prayer for Owen Meany          413   
193                            A Prayer for Owen Meany          413   
194   The Vampire Lestat (Vampire Chronicles, Book II)          301   
195       White Oleander : A Novel (Oprah's Book Club)          356   

     avg_ratings        ISBN         Book-Author Year-Of-Publication  \
0       5.852804  0439136350       J. K. Rowling                1999   
1       5.852804  0439136369       J. K. Rowling                2001   
2       5.852804  0786222743       J. K. Rowling                2000   
3       5.824289  0439139597       J. K. Rowling                2000   
4       5.824289  0439139600       J. K. Rowling                2002   
..           ...         ...                 ...                 ...   
191     3.818533  0698102193  William G. Golding                1978   
192     3.796610  0345361792         John Irving                1990   
193     3.796610  0345363523         John Irving                1989   
194     3.777409  0345313860           ANNE RICE                1986   
195     3.772472  0316284955         Janet Fitch                2000   

                 Publisher                                        Image-URL-S  \
0               Scholastic  http://images.amazon.com/images/P/0439136350.0...   
1               Scholastic  http://images.amazon.com/images/P/0439136369.0...   
2          Thorndike Press  http://images.amazon.com/images/P/0786222743.0...   
3               Scholastic  http://images.amazon.com/images/P/0439139597.0...   
4    Scholastic Paperbacks  http://images.amazon.com/images/P/0439139600.0...   
..                     ...                                                ...   
191       Putnam Pub Group  http://images.amazon.com/images/P/0698102193.0...   
192       Ballantine Books  http://images.amazon.com/images/P/0345361792.0...   
193       Ballantine Books  http://images.amazon.com/images/P/0345363523.0...   
194       Ballantine Books  http://images.amazon.com/images/P/0345313860.0...   
195         Back Bay Books  http://images.amazon.com/images/P/0316284955.0...   

                                           Image-URL-M  \
0    http://images.amazon.com/images/P/0439136350.0...   
1    http://images.amazon.com/images/P/0439136369.0...   
2    http://images.amazon.com/images/P/0786222743.0...   
3    http://images.amazon.com/images/P/0439139597.0...   
4    http://images.amazon.com/images/P/0439139600.0...   
..                                                 ...   
191  http://images.amazon.com/images/P/0698102193.0...   
192  http://images.amazon.com/images/P/0345361792.0...   
193  http://images.amazon.com/images/P/0345363523.0...   
194  http://images.amazon.com/images/P/0345313860.0...   
195  http://images.amazon.com/images/P/0316284955.0...   

                                           Image-URL-L  
0    http://images.amazon.com/images/P/0439136350.0...  
1    http://images.amazon.com/images/P/0439136369.0...  
2    http://images.amazon.com/images/P/0786222743.0...  
3    http://images.amazon.com/images/P/0439139597.0...  
4    http://images.amazon.com/images/P/0439139600.0...  
..                                                 ...  
191  http://images.amazon.com/images/P/0698102193.0...  
192  http://images.amazon.com/images/P/0345361792.0...  
193  http://images.amazon.com/images/P/0345363523.0...  
194  http://images.amazon.com/images/P/0345313860.0...  
195  http://images.amazon.com

In [109]:
# we see that the number of rows increased from 50 to 196, bcoz same book is having diff ISBN number.
# so we need to drop the duplicates on Book_Title.
popular_df.merge(books, on ='Book-Title').drop_duplicates('Book-Title').shape

(50, 10)

In [110]:
# since same book appeared in multiple columns due to different ISBN number, so we 
# removed the duplicates based on 'Book-Title.'

In [111]:
# taking only the columns that we need in
popular_df = popular_df.merge(books, on ='Book-Title').drop_duplicates('Book-Title')[['Book-Title','Book-Author','Image-URL-M','num_ratings','avg_ratings']]

In [112]:
popular_df.shape

(50, 5)

In [113]:
popular_df

Book-Title  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)   
3         Harry Potter and the Goblet of Fire (Book 4)   
5       Harry Potter and the Sorcerer's Stone (Book 1)   
9    Harry Potter and the Order of the Phoenix (Boo...   
13    Harry Potter and the Chamber of Secrets (Book 2)   
16   The Hobbit : The Enchanting Prelude to The Lor...   
17   The Fellowship of the Ring (The Lord of the Ri...   
26   Harry Potter and the Sorcerer's Stone (Harry P...   
28      The Two Towers (The Lord of the Rings, Part 2)   
39                               To Kill a Mockingbird   
47                                   The Da Vinci Code   
53                  The Five People You Meet in Heaven   
55                              The Catcher in the Rye   
62                           The Lovely Bones: A Novel   
63                                                1984   
72                            Prodigal Summer: A Novel   
73                                          Neverwhere   
78                             The Secret Life of Bees   
84   Stupid White Men ...and Other Sorry Excuses fo...   
85   Tuesdays with Morrie: An Old Man, a Young Man,...   
88                 The Red Tent (Bestselling Backlist)   
89                The Hitchhiker's Guide to the Galaxy   
98    Nickel and Dimed: On (Not) Getting By in America   
100                                      Into the Wild   
103                                     Fahrenheit 451   
116                          Girl with a Pearl Earring   
117                                          Outlander   
123  Where the Heart Is (Oprah's Book Club (Paperba...   
124                     Seabiscuit: An American Legend   
126                                         Life of Pi   
130                                     The Bean Trees   
133  A Child Called \It\": One Child's Courage to S...   
135                                     ANGELA'S ASHES   
137                                        Good in Bed   
139  Fast Food Nation: The Dark Side of the All-Ame...   
140                                      American Gods   
141                                 Skipping Christmas   
147  Left Behind: A Novel of the Earth's Last Days ...   
149    The Golden Compass (His Dark Materials, Book 1)   
157                                   The Color Purple   
163                               Silence of the Lambs   
166                                        About a Boy   
173                  Seven Up (A Stephanie Plum Novel)   
175  The Alchemist: A Fable About Following Your Dream   
177  Hard Eight : A Stephanie Plum Novel (A Stephan...   
182                       Suzanne's Diary for Nicholas   
186                                  Lord of the Flies   
192                            A Prayer for Owen Meany   
194   The Vampire Lestat (Vampire Chronicles, Book II)   
195       White Oleander : A Novel (Oprah's Book Club)   

                Book-Author  \
0             J. K. Rowling   
3             J. K. Rowling   
5             J. K. Rowling   
9             J. K. Rowling   
13            J. K. Rowling   
16           J.R.R. TOLKIEN   
17           J.R.R. TOLKIEN   
26            J. K. Rowling   
28           J.R.R. TOLKIEN   
39               Harper Lee   
47                Dan Brown   
53              Mitch Albom   
55            J.D. Salinger   
62             Alice Sebold   
63            George Orwell   
72       Barbara Kingsolver   
73              Neil Gaiman   
78            Sue Monk Kidd   
84            Michael Moore   
85              MITCH ALBOM   
88            Anita Diamant   
89            Douglas Adams   
98       Barbara Ehrenreich   
100            Jon Krakauer   
103            Ray Bradbury   
116         Tracy Chevalier   
117          DIANA GABALDON   
123            Billie Letts   
124       LAURA HILLENBRAND   
126             Yann Martel   
130      Barbara Kingsolver   
133             Dave Pelzer   
135           Frank McCourt   
137         Jennifer Weiner   


# Collaborative Filtering Based Recommender system

In [114]:
# we will consider only those users who voted on > 250 books (experienced users)
# Also the books having > 50  votes will only be considered.

In [115]:
# applied groupby on 'User-ID', then counted the Book-Ratings given by each user.
ratings_with_book.groupby('User-ID').count()['Book-Rating']

User-ID
2          1
8         17
9          3
10         1
12         1
          ..
278846     1
278849     4
278851    23
278852     1
278854     8
Name: Book-Rating, Length: 92106, dtype: int64

In [116]:
users.shape
# so out of 2,78,858 users only 92,106 rated  any book.

(278858, 3)

In [117]:
# storing the indices of the users who voted on +200 books.
x = ratings_with_book.groupby('User-ID').count()['Book-Rating'] > 200
padhe_likhe_users = x[x].index

In [118]:
x[x].index

Int64Index([   254,   2276,   2766,   2977,   3363,   4017,   4385,   6251,
              6323,   6543,
            ...
            271705, 273979, 274004, 274061, 274301, 274308, 275970, 277427,
            277639, 278418],
           dtype='int64', name='User-ID', length=811)

In [119]:
# now filtering these indices in the ratings_with_book data.
filtered_rating = ratings_with_book[ ratings_with_book['User-ID'].isin(padhe_likhe_users)]

In [120]:
filtered_rating

User-ID        ISBN  Book-Rating  \
2           6543  034545104X            0   
5          23768  034545104X            0   
7          28523  034545104X            0   
15         77940  034545104X            0   
16         81977  034545104X            0   
...          ...         ...          ...   
1030883   275970  1880837927            0   
1030884   275970  188717897X            0   
1030885   275970  1888889047            0   
1030886   275970  1931868123            0   
1030887   275970  3411086211           10   

                                                Book-Title  \
2                                     Flesh Tones: A Novel   
5                                     Flesh Tones: A Novel   
7                                     Flesh Tones: A Novel   
15                                    Flesh Tones: A Novel   
16                                    Flesh Tones: A Novel   
...                                                    ...   
1030883                         The Theology of the Hammer   
1030884  The Ordeal of Integration: Progress and Resent...   
1030885  Pushcart's Complete Rotten Reviews &amp; Rejec...   
1030886  There's a Porcupine in My Outhouse: Misadventu...   
1030887                                         Die Biene.   

                           Book-Author Year-Of-Publication  \
2                           M. J. Rose                2002   
5                           M. J. Rose                2002   
7                           M. J. Rose                2002   
15                          M. J. Rose                2002   
16                          M. J. Rose                2002   
...                                ...                 ...   
1030883                 Millard Fuller                1994   
1030884              Orlando Patterson                1998   
1030885                 Bill Henderson                1998   
1030886                   Mike Tougias                2002   
1030887  Sybil Gr�¤fin Sch�¶nfeldt                1993   

                                    Publisher  \
2                            Ballantine Books   
5                            Ballantine Books   
7                            Ballantine Books   
15                           Ballantine Books   
16                           Ballantine Books   
...                                       ...   
1030883         Smyth &amp; Helwys Publishing   
1030884                Civitas Book Publisher   
1030885                        Pushcart Press   
1030886                    Capital Books (VA)   
1030887  Bibliographisches Institut, Mannheim   

                                               Image-URL-S  \
2        http://images.amazon.com/images/P/034545104X.0...   
5        http://images.amazon.com/images/P/034545104X.0...   
7        http://images.amazon.com/images/P/034545104X.0...   
15       http://images.amazon.com/images/P/034545104X.0...   
16       http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1030883  http://images.amazon.com/images/P/1880837927.0...   
1030884  http://images.amazon.com/images/P/188717897X.0...   
1030885  http://images.amazon.com/images/P/1888889047.0...   
1030886  http://images.amazon.com/images/P/1931868123.0...   
1030887  http://images.amazon.com/images/P/3411086211.0...   

                                               Image-URL-M  \
2        http://images.amazon.com/images/P/034545104X.0...   
5        http://images.amazon.com/images/P/034545104X.0...   
7        http://images.amazon.com/images/P/034545104X.0...   
15       http://images.amazon.com/images/P/034545104X.0...   
16       http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1030883  http://images.amazon.com/images/P/1880837927.0...   
1030884  http://images.amazon.com/images/P/188717897X.0...   
1030885  http://images.amazon.com/images/P/1888889047.0...   
1030886  http://images.a

In [121]:
# no of rows: 4.74 lakh
# total ratings were 10 lakh--> 4.74 lakh ratings are done by padhe-likhe-users.

In [122]:
# we have prepared the dataset for ratings & their user.
# now we will fiter on Book-Title and Book-Rating.
y = filtered_rating.groupby('Book-Title').count()['Book-Rating'] >=50

In [123]:
# y gives the boolean indexing of books which received users=ratings > 50
# y[y] gives all the true indexed data.
famous_book = y[y].index

In [124]:
# getting all the rows which comes under famous books.
final_ratings = filtered_rating[filtered_rating['Book-Title'].isin(famous_book)]

In [125]:
final_ratings

User-ID        ISBN  Book-Rating              Book-Title  \
63        278418  0446520802            0            The Notebook   
65          3363  0446520802            0            The Notebook   
66          7158  0446520802           10            The Notebook   
69         11676  0446520802           10            The Notebook   
74         23768  0446520802            6            The Notebook   
...          ...         ...          ...                     ...   
1026724   266865  0531001725           10  The Catcher in the Rye   
1027923   269566  0670809381            0                  Echoes   
1028777   271284  0440910927            0           The Rainmaker   
1029070   271705  B0001PIOX4            0          Fahrenheit 451   
1030868   275970  1586210661            9  Me Talk Pretty One Day   

                   Book-Author Year-Of-Publication                Publisher  \
63             Nicholas Sparks                1996             Warner Books   
65             Nicholas Sparks                1996             Warner Books   
66             Nicholas Sparks                1996             Warner Books   
69             Nicholas Sparks                1996             Warner Books   
74             Nicholas Sparks                1996             Warner Books   
...                        ...                 ...                      ...   
1026724  Jerome David Salinger                1973   Scholastic Library Pub   
1027923           Maeve Binchy                1986              Penguin USA   
1028777           John Grisham                1995                   Island   
1029070           Ray Bradbury                1993     Simon &amp; Schuster   
1030868          David Sedaris                2001  Time Warner Audio Major   

                                               Image-URL-S  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-M  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-L  
63       http://images.amazon.com/images/P/0446520802.0...  
65       http://images.amazon.com/images/P/0446520802.0...  
66       http://images.amazon.com/images/P/0446520802.0...  
69       http://images.amazon.com/images/P/0446520802.0...  
74       http://images.amazon.com/images/P/0446520802.0...  
...                                                    ...  
1026724  http://images.amazon.com/images/P/0531001725.0...  
1027923  http://images.amazon.com/images/P/0670809381.0...  
1028777  http://images.amazon.com/images/P/0440910927.0...  
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...  
1030868  http://images.amazon.com/images/P/1586210661.0...  

[58586 r

In [126]:
final_ratings.index

Int64Index([     63,      65,      66,      69,      74,      77,      78,
                 79,      81,      82,
            ...
            1025261, 1025321, 1025418, 1025422, 1025428, 1026724, 1027923,
            1028777, 1029070, 1030868],
           dtype='int64', length=58586)

In [127]:
# creating a pivot-table ( each user vs their rating on a book)
pivot_table=final_ratings.pivot_table(index='Book-Title',columns='User-ID',values='Book-Rating')

In [128]:
pivot_table.fillna(0,inplace=True)
pivot_table

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

In [129]:
# ---> 706 books are famous and 810 users are padhe-likhe

In [130]:
# let's visualize each book in 810 dimensions, because it recieved ratings from those users.
#### Book1 --> [., ., .,., ., ., ., ., ...........,810th user]
# thus each book is a vector in 810 dimensional space.

In [131]:
# now we will find the distance of each book from other books---> using cosine similarity.
# it is used to find the amount of similarity b/w two data points.


In [132]:
from sklearn.metrics.pairwise import cosine_similarity
similarity_score=cosine_similarity(pivot_table)

In [133]:
similarity_score.shape
# 706 such distances of each 706 books have been stored in similarity_score

(706, 706)

In [134]:
# checking the location of a book, in pivot_table.
# pivot_table.index=='Zoya' is 1*706 matrix.
np.where(pivot_table.index=='Zoya')[0][0]

704

In [135]:
# enumerating on the list of similarity score of book1 in reverse-sorted order. 
# x[1]--> 1 because, similarity score is on 1st index and on 0th index its simply the index-number.
# and finally getting 1-6 items, since 0th is the book itself so its score is 0.99999.
sorted(list(enumerate(similarity_score[0])), key=lambda x: x[1],reverse=True)[1:6]

[(47, 0.2702651417103732),
 (545, 0.2639619371123497),
 (82, 0.23669374347400993),
 (634, 0.23299389358170394),
 (551, 0.2262639743141286)]

In [174]:
# Now will make a function which will return top 'n' books similar to a given book.(based on most similar scores.)
def recommend(book_name):
    #fetch the index of book-name
    index = np.where(pivot_table.index==book_name)[0][0]
    similar_items = sorted(list(enumerate(similarity_score[index])), key=lambda x: x[1],reverse=True)[1:6]
    
    # now we got the similar_items(which has book-index and score), so will loop into it to print them.
    # using append makes the list 2-D list
    data=[]
    for i in similar_items:
        item =[]
        temp_df = books[books['Book-Title'] == pivot_table.index[i[0]]]
        item.extend(temp_df.drop_duplicates('Book-Title')['Book-Title'].values)
        item.extend(temp_df.drop_duplicates('Book-Title')['Book-Author'].values)
        item.extend(temp_df.drop_duplicates('Book-Title')['Image-URL-M'].values)
        
        data.append(item)
    return data


In [175]:
recommend('1984')

[['Animal Farm',
  'George Orwell',
  'http://images.amazon.com/images/P/0451526341.01.MZZZZZZZ.jpg'],
 ["The Handmaid's Tale",
  'Margaret Atwood',
  'http://images.amazon.com/images/P/0449212602.01.MZZZZZZZ.jpg'],
 ['Brave New World',
  'Aldous Huxley',
  'http://images.amazon.com/images/P/0060809833.01.MZZZZZZZ.jpg'],
 ['The Vampire Lestat (Vampire Chronicles, Book II)',
  'ANNE RICE',
  'http://images.amazon.com/images/P/0345313860.01.MZZZZZZZ.jpg'],
 ['The Hours : A Novel',
  'Michael Cunningham',
  'http://images.amazon.com/images/P/0312243022.01.MZZZZZZZ.jpg']]

In [140]:
     
 ##  recommend('1984') gives the indices of top 5 similar books.
 # pivot_table.index gives the list of all the books. but we want only 5 books


In [172]:
# since the recommend function is depending on pivot_table, similarity_score and books,
# so we need to export them as pkl file too..
pickle.dump(books,open('books.pkl','wb'))
pickle.dump(similarity_score,open('similarity_score.pkl','wb'))
pickle.dump(pivot_table,open('pivot_table.pkl','wb'))

In [173]:
data

NameError: name 'data' is not defined

In [142]:
import pickle
pickle.dump(popular_df, open('popular.pkl','wb'))